linkerd/distribute/src/service.rs

```diff
@@ -0,0 +1,317 @@
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = S::Future;
+
+    /// Acquires a ready backend.
+    ///
+    /// Note that this doesn't necessarily drive all backend services to
+    /// readiness. We expect that these inner services should be buffered or
+    /// otherwise drive themselves to readiness (i.e. via SpawnReady).
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        // If we've already chosen a ready index, then skip polling.
+        if self.ready_idx.is_some() {
+            return Poll::Ready(Ok(()));
+        }
+
+        match self.selection {
+            Selection::Empty => {
+                tracing::debug!("empty distribution will never become ready");
+            }
+
+            Selection::FirstAvailable => {
+                for (idx, svc) in self.backends.values_mut().enumerate() {
+                    if svc.poll_ready(cx)?.is_ready() {
+                        self.ready_idx = Some(idx);
+                        return Poll::Ready(Ok(()));
+                    }
+                }
+            }
+
+            // Choose a random index (via the weighted distribution) to try to
+            // poll the backend. Continue selecting endpoints until we find one
+            // that is ready or we've tried all backends in the distribution.
+            Selection::RandomAvailable {
+                ref keys,
+                ref mut rng,
+            } => {
+                // Clone the weighted index so that we can zero out the weights
+                // as pending services are polled.
+                let mut index = keys.index().clone();
+                loop {
+                    // Sample the weighted index to find a backend to try.
+                    let idx = index.sample(rng);
+                    let (_, svc) = self
+                        .backends
+                        .get_index_mut(idx)
+                        .expect("distributions must not reference unknown backends");
+
+                    if svc.poll_ready(cx)?.is_ready() {
+                        self.ready_idx = Some(idx);
+                        return Poll::Ready(Ok(()));
+                    }
+
+                    // Zero out the weight of the backend we just tried so that
+                    // it's not selected again.
+                    match index.update_weights(&[(idx, &0)]) {
+                        Ok(()) => {}
+                        Err(WeightedError::AllWeightsZero) => {
+                            // There are no backends remaining.
+                            break;
+                        }
+                        Err(error) => {
+                            tracing::error!(%error, "unexpected error updating weights; giving up");
+                            break;
+                        }
+                    }
+                }
+            }
+        }
+
+        debug_assert!(self.ready_idx.is_none());
+        Poll::Pending
```

Comment:
nit: should we log something like
```suggestion
        tracing::trace!("no ready services in distribution");
        Poll::Pending
```
here?